numpy/core/src/multiarray/mapping.c

```diff
@@ -756,6 +760,63 @@ prepare_index(PyArrayObject *self, PyObject *index,
                          NPY_MAXDIMS, (new_ndim + fancy_ndim));
             goto failed_building_indices;
         }
+
+        /*
+         * If we had a fancy index, we may have had a boolean array index.
+         * So check if this had the correct shape, now that we can find out
+         * which axes it acts on.
+         */
+        used_ndim = 0;
+        for (i=0; i < curr_idx; i++) {
+            if ((indices[i].type == HAS_FANCY) && indices[i].value > 0) {
+                if (indices[i].value != PyArray_DIM(self, used_ndim)) {
+                    PyObject *m, *warning;
+
+                    char *err_msg[174];
+                    sprintf(err_msg,
+                        "boolean index did not match indexed array along "
+                        "dimension %d; dimension is %" NPY_INTP_FMT
+                        " but corresponding boolean dimension is %" NPY_INTP_FMT,
+                        used_ndim, PyArray_DIM(self, used_ndim),
+                        indices[i].value);
+
+                    m = PyImport_ImportModule("numpy");
+                    if (m == NULL) {
+                        PyErr_SetString(
+                            PyExc_SystemError,
```

Comment:
Might as well put `PyExc_SystemError` on previous line. There is a (secret) style exception for that.